src/www/app/outline_server.spec.ts

```diff
@@ -92,6 +92,216 @@ describe('OutlineServerRepository', () => {
       accessKey: shadowsocksConfigToAccessKey(CONFIG_1)
     });
   });
+
+  it('add stores servers', () => {
+    const storage = new InMemoryStorage();
+    const repo = new OutlineServerRepository(getFakeServerFactory(), new EventQueue(), storage);
+    const accessKey0 = shadowsocksConfigToAccessKey(CONFIG_0);
+    const accessKey1 = shadowsocksConfigToAccessKey(CONFIG_1);
+    repo.add(accessKey0);
+    repo.add(accessKey1);
+    const servers: ServersStorageV1 =
+        JSON.parse(storage.getItem(OutlineServerRepository.SERVERS_STORAGE_KEY));
+    expect(servers.length).toEqual(2);
+    expect(servers[0].accessKey).toEqual(accessKey0);
+    expect(servers[0].name).toEqual(CONFIG_0.name);
+    expect(servers[1].accessKey).toEqual(accessKey1);
+    expect(servers[1].name).toEqual(CONFIG_1.name);
+  });
+
+  it('add emits ServerAdded event', async () => {
+    const eventQueue = new EventQueue();
+    const repo =
+        new OutlineServerRepository(getFakeServerFactory(), eventQueue, new InMemoryStorage());
+    const accessKey = shadowsocksConfigToAccessKey(CONFIG_0);
+    repo.add(accessKey);
+    await new Promise<void>(resolve => {
```

Comment:
I don't think this catches the failure of the `ServerAdded` event not getting fired. Does the test fail if the promise doesn't resolve in time?

How about instead we await on `eventQueue.startPublishing()` (seems like it's synchronous) and `fail()` if we don't receive the event?

Reply:
jasmine fails tests that don't complete after 3s, so this works as expected. I made these tests synchronous per your suggestion in 54492e2.